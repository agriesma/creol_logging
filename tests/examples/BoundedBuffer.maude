load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op state : -> State [ctor] .
  eq state = {
    < "BoundedBuffer" : Class | Inh: noInh, Param: noVid, Att: "buffer" |->
      null, "max" |-> null, "n" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( ( "buffer" @ "BoundedBuffer" ),
              ( "max" @ "BoundedBuffer" ),
              ( "n" @ "BoundedBuffer" ) ; list(emp) :: int(10) :: int(0) ) ;
              skip ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "append" : Method | Param: "d", Att: "d" |-> null, "_" |->
              null, Code: await "<" ( "n" :: "max" ) ; assign( "buffer",
              "n" ; "|-" ( "buffer" :: "d" ) :: "+" ( "n" :: int(1) ) ) ;
              return ( emp ) >,
            < "remove" : Method | Param: noVid, Att: "d" |-> null, "_" |->
              null, Code: await ">" ( "n" :: int(0) ) ;
              assign( "d" ; "head" ( "buffer" ) ) ; assign( "buffer",
              "n" ; "tail" ( "buffer" ) :: "-" ( "n" :: int(1) ) ) ;
              return ( "d" ) >,
      Ocnt: 0 >
    
    < "Producer" : Class | Inh: noInh, Param: "b", Att: "b" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "ccglab0" |-> null, "_" |->
              null, Code: call( "ccglab0" ; "this" ; "loop" ; int(0) ) ;
              free( "ccglab0" ) ; return ( emp ) >,
            < "loop" : Method | Param: "i", Att: "i" |-> null, "ccglab1" |->
              null, "ccglab0" |-> null, "_" |-> null, Code: call( "ccglab0" ;
              "b" ; "append" ; "i" ) ; get( "ccglab0" ; noVid ) ;
              free( "ccglab0" ) ;
              call( "ccglab1" ; "this" ; "loop" ; "+" ( "i" :: int(1) ) ) ;
              free( "ccglab1" ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < "Consumer" : Class | Inh: noInh, Param: "b", Att: "b" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "ccglab1" |-> null,
              "ccglab0" |-> null, "y" |-> null, "_" |-> null,
              Code: call( "ccglab0" ; "b" ; "remove" ; emp ) ;
              get( "ccglab0" ; "y" ) ; free( "ccglab0" ) ;
              call( "ccglab1" ; "this" ; "run" ; emp ) ; free( "ccglab1" ) ;
              return ( emp ) >,
      Ocnt: 0 >
    
    < "Starter" : Class | Inh: noInh, Param: noVid, Att: noSubst,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "b" |-> null, "p" |-> null,
              "c" |-> null, "_" |-> null,
              Code: new( "b" ; "BoundedBuffer" ; emp ) ;
              new( "p" ; "Producer" ; "b" ) ; new( "c" ; "Consumer" ; "b" ) ;
              return ( emp ) >,
      Ocnt: 0 > } .
endm

