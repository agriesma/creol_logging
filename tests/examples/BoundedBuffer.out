==========================================
reduce in PROGRAM : state .
result State: {
< "BoundedBuffer" : Class | Inh: noInh, Param: noVid, Att: "buffer" |-> null,
    "max" |-> null, "n" |-> null, Mtds:
  < "append" : Method | Param: "d", Att: "_" |-> null, "d" |-> null, Code:
    await "<"("n" :: "max") ; assign("buffer", "n" ; "|-"("buffer" :: "d") ::
    "+"("n" :: int(1))) ; return(emp) >, < "init" : Method | Param: noVid, Att:
    "_" |-> null, Code: assign(("buffer" @ "BoundedBuffer"), ("max" @
    "BoundedBuffer"), ("n" @ "BoundedBuffer") ; list(emp) :: int(10) :: int(0))
    ; skip ; return(emp) >, < "remove" : Method | Param: noVid, Att: "_" |->
    null, "d" |-> null, Code: await ">"("n" :: int(0)) ; assign("d" ; "head"(
    "buffer")) ; assign("buffer", "n" ; "tail"("buffer") :: "-"("n" :: int(1)))
    ; return("d") >, < "run" : Method | Param: noVid, Att: "_" |-> null, Code:
    skip ; return(emp) >,
  Ocnt: 0 >
 
< "Consumer" : Class | Inh: noInh, Param: "b", Att: "b" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "ccglab0" |-> null,
    "ccglab1" |-> null, "y" |-> null, Code: call("ccglab0" ; "b" ; "remove" ;
    emp) ; get("ccglab0" ; "y") ; free("ccglab0") ; call("ccglab1" ; "this" ;
    "run" ; emp) ; free("ccglab1") ; return(emp) >,
  Ocnt: 0 >
 
< "Producer" : Class | Inh: noInh, Param: "b", Att: "b" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "loop" : Method | Param: "i", Att: "_" |-> null, "ccglab0" |-> null,
    "ccglab1" |-> null, "i" |-> null, Code: call("ccglab0" ; "b" ; "append" ;
    "i") ; get("ccglab0" ; noVid) ; free("ccglab0") ; call("ccglab1" ; "this" ;
    "loop" ; "+"("i" :: int(1))) ; free("ccglab1") ; return(emp) >, < "run" :
    Method | Param: noVid, Att: "_" |-> null, "ccglab0" |-> null, Code: call(
    "ccglab0" ; "this" ; "loop" ; int(0)) ; free("ccglab0") ; return(emp) >,
  Ocnt: 0 >
 
< "Starter" : Class | Inh: noInh, Param: noVid, Att: noSubst, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "b" |-> null, "c"
    |-> null, "p" |-> null, Code: new("b" ; "BoundedBuffer" ; emp) ; new("p" ;
    "Producer" ; "b") ; new("c" ; "Consumer" ; "b") ; return(emp) >,
  Ocnt: 0 >
}
Bye.
